fem/hyperbolic.cpp

```diff
@@ -212,15 +448,290 @@ real_t RusanovFlux::Eval(const Vector &state1, const Vector &state2,
+      {
+         flux(i) = std::min(fluxN1(i), fluxN2(i));
+      }
+      else
+      {
+         flux(i) = std::max(fluxN1(i), fluxN2(i));
+      }
    }
+
    return std::max(speed1, speed2);
 }
 
+void GodunovFlux::Grad(int side, const Vector &state1, const Vector &state2,
+                       const Vector &nor, FaceElementTransformations &Tr,
+                       DenseMatrix &grad) const
+{
+#ifdef MFEM_THREAD_SAFE
+   DenseMatrix JDotN(fluxFunction.num_equations);
+#else
+   JDotN.SetSize(fluxFunction.num_equations);
+#endif
+
+   grad = 0.;
+
+   if (side == 1)
+   {
+      fluxFunction.ComputeFluxJacobianDotN(state1, nor, Tr, JDotN);
+
+      for (int i = 0; i < fluxFunction.num_equations; i++)
+      {
+         // Only diagonal terms of J are considered
+         grad(i,i) = std::max(JDotN(i,i), 0_r);
+      }
+   }
+   else
+   {
+      fluxFunction.ComputeFluxJacobianDotN(state2, nor, Tr, JDotN);
+
+      for (int i = 0; i < fluxFunction.num_equations; i++)
+      {
+         // Only diagonal terms of J are considered
+         grad(i,i) = std::min(JDotN(i,i), 0_r);
+      }
+   }
+}
+
+real_t GodunovFlux::Average(const Vector &state1, const Vector &state2,
+                            const Vector &nor, FaceElementTransformations &Tr,
+                            Vector &flux) const
+{
+#ifdef MFEM_THREAD_SAFE
+   Vector fluxN1(fluxFunction.num_equations), fluxN2(fluxFunction.num_equations);
+#endif
+   const real_t speed1 = fluxFunction.ComputeFluxDotN(state1, nor, Tr, fluxN1);
+   const real_t speed2 = fluxFunction.ComputeAvgFluxDotN(state1, state2, nor, Tr,
+                                                         fluxN2);
+
+   for (int i = 0; i < fluxFunction.num_equations; i++)
+   {
+      if (state1(i) <= state2(i))
+      {
+         flux(i) = std::min(fluxN1(i), fluxN2(i));
+      }
+      else
+      {
+         flux(i) = std::max(fluxN1(i), fluxN2(i));
+      }
+   }
+
+   return std::max(speed1, speed2);
+}
+
+void GodunovFlux::AverageGrad(int side, const Vector &state1,
+                              const Vector &state2,
+                              const Vector &nor, FaceElementTransformations &Tr,
+                              DenseMatrix &grad) const
+{
+#ifdef MFEM_THREAD_SAFE
+   Vector fluxN1(fluxFunction.num_equations), fluxN2(fluxFunction.num_equations);
+#endif
+   if (side == 1)
+   {
+#ifdef MFEM_THREAD_SAFE
+      DenseMatrix JDotN(fluxFunction.num_equations);
+#else
+      JDotN.SetSize(fluxFunction.num_equations);
+#endif
+      fluxFunction.ComputeFluxDotN(state1, nor, Tr, fluxN1);
+      fluxFunction.ComputeAvgFluxDotN(state1, state2, nor, Tr, fluxN2);
+      fluxFunction.ComputeFluxJacobianDotN(state1, nor, Tr, JDotN);
+
+      grad = 0.;
+
+      for (int i = 0; i < fluxFunction.num_equations; i++)
+      {
+         // Only diagonal terms of J are considered
+         // lim_{u → u⁻} (F̄(u⁻,u)n - F(u⁻)n) / (u - u⁻) = ½J(u⁻)n
+         const real_t gr12 = (state1(i) != state2(i))?
```

Comment:
Do you need a numerical tolerance here? The reciprocal term might be an issue

Reply:
This is mainly for the IC or initial guess, where you have exactly the same values. Maybe one would need some tolerance for propagation of some infinitesimal waves, but "physical" infinitesimal waves are usually well above the floating point precision, so it might be rather a concern for propagation of some spurious waves and stability of the solution. It is just a simple finite difference, so I do not think it is a big issue, but I am open to adding something. Do you have a suggestion for some robust definition of such tolerance?

Reply:
I'm more worried about u2-u1 being on the order of machine precision (but nonzero), it could return values that are unfeasibly large. If you're sure that it's always exactly the same, that's fine (although maybe add a comment). Otherwise, you could just do the switch based on `abs(state1(i) - state2(i)) > eps` instead of `state1(i) != state2(i)` and have `eps` be a few orders of magnitude larger than machine precision (same for the Rusanov flux).